runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -1184,4 +1184,104 @@ function LanguageTree:language_for_range(range)
   return self
 end
 
+--- Get highlight information from parser.
+---
+---@return {group:string, range:integer[]}
+function LanguageTree:get_highlight()
+  ---@type {group:string,range:integer[],priority:integer,type:'start'|'end',start?:table,subpriority:integer}[]
+  local positions = {}
+  self:for_each_tree(function(tree, ltree)
+    local source = ltree:source()
+    local hl_query = query.get(ltree:lang(), 'highlights')
+
+    if not hl_query then
+      return
+    end
+
+    for id, node, metadata in hl_query:iter_captures(tree:root(), source, 0, -1) do
+      local name = hl_query.captures[id]
+      local start_row, start_col, end_row, end_col = node:range()
+      local priority = tonumber(metadata.priority or vim.highlight.priorities.treesitter)
+
+      table.insert(positions, {
+        group = '@' .. name .. '.' .. ltree:lang(),
+        range = { start_row, start_col },
+        priority = priority,
+        subpriority = #positions,
+        type = 'start',
+      })
+
+      table.insert(positions, {
+        group = '@' .. name .. '.' .. ltree:lang(),
+        range = { end_row, end_col },
+        priority = priority,
+        subpriority = #positions - 1,
+        type = 'end',
+      })
+    end
+  end)
+
+  table.sort(positions, function(a, b)
+    return a.range[1] < b.range[1]
+      or (a.range[1] == b.range[1] and a.range[2] < b.range[2])
+      or (
+        a.range[1] == b.range[1]
+        and a.range[2] == b.range[2]
+        and a.type == 'start'
+        and b.type == 'end'
+      )
+  end)
+
+  local result = {}
+  ---@type  {group:string,range:integer[],priority:integer,type:'start'|'end',start?:table,subpriority:integer}[]
```

Comment:
Since this has 6 fields, it would be much more readable as a `@class`.

`binary_find` can also be pulled out as a put function with `stack` as an argument.